modules/oe_whitelabel_helper/src/Plugin/field_group/FieldGroupFormatter/PreRenderingFieldGroupFormatterBase.php

```diff
@@ -0,0 +1,75 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\oe_whitelabel_helper\Plugin\field_group\FieldGroupFormatter;
+
+use Drupal\Core\Render\Element;
+use Drupal\Core\Security\Attribute\TrustedCallback;
+use Drupal\field_group\FieldGroupFormatterBase;
+
+/**
+ * Special base class for field group formatters.
+ *
+ * This is meant for field group formatters that want their logic to run in the
+ * '#pre_render' step, when the final list of visible fields has been correctly
+ * determined.
```

Comment:
### Emptiness check
Field group has a built-in mechanism to detect emptiness.
However, this mechanism only happens after the individual field group formatters already did their `preRender()` method. This means the `preRender()` method does not know if the field group is empty or not.

The built-in field group formatters like Accordion or Fieldset use dedicated element types and '#theme_wrappers' which expect the content to be direct children of the root element of the group, and in the end nothing is printed if the group has no content.

In the existing pattern-based field group formatters in `oe_whitelabel`, we actually render the content to determine whether it is empty. E.g. in `DescriptionListPattern`, we do this:

```php
      $field_markup = $this->renderer->render($field_element);
      if (trim((string) $field_markup) === '') {
        continue;
      }
```

This probably works most of the time, but it means we are replacing an existing default solution with a custom solution.

### Flexibility and DX
With the `PatternFormatterBase`, a subclass has to implement two methods, `getPatternId()` and `getFields()`, where the `getFields()` has to do the emptiness check in the subclass. The `DescriptionListPattern` even needs to override the `preRender()` method itself, to customize pattern settings.

The `PreRenderingFieldGroupFormatterBase` reduces the complexity of the extending class, while at the same time allowing usage beyond patterns.

The `PreRenderingPatternFormatterBase` brings back the mechanism to choose a pattern variant.

### Can it fully replace the old base class?
I think yes, in theory. But we have to keep the old class around for BC, and even the exiting subclasses we can't simply rewrite, because that would also have a BC impact.

(I think I was somehow involved in the design of the old base class, although I see Sergio as the commit author. So maybe not as much.)

Obviously this would cause a situation where we have different alternative base classes, which can cause confusion.